File: cpp/test/Test_Matrix.cpp
#include "Test.h"
#include "Matrix.h"
#include <cmath>
#include <iostream>
using namespace std;
using namespace zmat;

void Test_Matrix_Equal() {
    Matrix a(3, 3);
    Matrix b(3, 3);
    Matrix c(3, 3);
    for (int i = 1; i <= 3; ++i) {
        for (int j = 1; j <= 3; ++j) {
            a(i, j) = i + j;
            b(i, j) = j + i;
            c(i, j) = i - j;
        }
    }
    TEST_EQUAL(a, b);
    TEST_EQUAL(b, a);
    TEST_UNEQUAL(a, c);
    TEST_UNEQUAL(c, a);
    TEST_UNEQUAL(b, c);
    TEST_UNEQUAL(c, b);
}

void Test_Matrix_Add() {
    Matrix a(3, 3);
    Matrix b(3, 3);
    Matrix c(3, 3);
    for (int i = 1; i <= 3; ++i) {
        for (int j = 1; j <= 3; ++j) {
            a(i, j) = i + j;
            b(i, j) = i - j;
            c(i, j) = i * 2;
        }
    }
    TEST_EQUAL(a + b, c);
    TEST_EQUAL(b + a, c);
}

void Test_Matrix_Subtract() {
    Matrix a(3, 3);
    Matrix b(3, 3);
    Matrix c(3, 3);
    for (int i = 1; i <= 3; ++i) {
        for (int j = 1; j <= 3; ++j) {
            a(i, j) = i + j;
            b(i, j) = i - j;
            c(i, j) = j * 2;
        }
    }
    TEST_EQUAL(a - b, c);
    TEST_UNEQUAL(b - a, c);
}

void Test_Matrix_Multiply() {
    Matrix a(3, 3);
    Matrix b(3, 3);
    Matrix c(3, 3);
    Matrix d(3, 3);
    for (int i = 1; i <= 3; ++i) {
        for (int j = 1; j <= 3; ++j) {
            a(i, j) = i + j;
            b(i, j) = i - j;
        }
    }
    c(1, 1) = 11;   c(1, 2) = 2;    c(1, 3) = -7;
    c(2, 1) = 14;   c(2, 2) = 2;    c(2, 3) = -10;
    c(3, 1) = 17;   c(3, 2) = 2;    c(3, 3) = -13;
    TEST_EQUAL(a * b, c);
    d(1, 1) = -11;  d(1, 2) = -14;  d(1, 3) = -17;
    d(2, 1) = -2;   d(2, 2) = -2;   d(2, 3) = -2;
    d(3, 1) = 7;    d(3, 2) = 10;   d(3, 3) = 13;
    TEST_EQUAL(b * a, d);
}

void Test_Matrix() {
    Test::add("Matrix", "Equal", Test_Matrix_Equal);
    Test::add("Matrix", "Add", Test_Matrix_Add);
    Test::add("Matrix", "Subtract", Test_Matrix_Subtract);
    Test::add("Matrix", "Multiply", Test_Matrix_Multiply);
}

File: js/test/test_matrix.js
test("Matrix", function() {
    var a = Matrix.New(3, 4);
    a.set(1, 2, Complex.New(2.0));
    ok(a.get(1, 2).equal(Complex.New(2.0)), "Set/Get");
    ok(a.get(2, 3).equal(Complex.New(0.0)), "Set/Get");
    var b = Matrix.New(3, 3);
    var c = Matrix.New(3, 3);
    for (var i = 1; i <= 3; ++i) {
        for (var j = 1; j <= 3; ++j) {
            b.set(i, j, ((i - 1) * 3 + j) * 2 - 1);
            c.set(i, j, ((i - 1) * 3 + j) * 2);
        }
    }
    a = Matrix.New(3, 3);
    a.set(1, 1, 3); a.set(1, 2, 7); a.set(1, 3, 11);
    a.set(2, 1, 15); a.set(2, 2, 19); a.set(2, 3, 23);
    a.set(3, 1, 27); a.set(3, 2, 31); a.set(3, 3, 35);
    ok(a.equal(b.add(c)), "Add");
    a.set(1, 1, -1); a.set(1, 2, -1); a.set(1, 3, -1);
    a.set(2, 1, -1); a.set(2, 2, -1); a.set(2, 3, -1);
    a.set(3, 1, -1); a.set(3, 2, -1); a.set(3, 3, -1);
    ok(a.equal(b.sub(c)), "Subtract");
    a.set(1, 1, 96); a.set(1, 2, 114); a.set(1, 3, 132);
    a.set(2, 1, 240); a.set(2, 2, 294); a.set(2, 3, 348);
    a.set(3, 1, 384); a.set(3, 2, 474); a.set(3, 3, 564);
    ok(a.equal(b.mul(c)), "Multiply 1");
    a.set(1, 1, 108); a.set(1, 2, 132); a.set(1, 3, 156);
    a.set(2, 1, 234); a.set(2, 2, 294); a.set(2, 3, 354);
    a.set(3, 1, 360); a.set(3, 2, 456); a.set(3, 3, 552);
    ok(a.equal(c.mul(b)), "Multiply 2");
});
